### ISSUE TITLE
Readonly Synced Realm Opens, But Closes Immediately

### ISSUE THREAD

**Issue by jarrodparkes** [realm/realm-cocoa#4485]
## Goals

What do you want to achieve?

- iOS app connects to Realm Object Server using an account with readonly permissions for the realm called `realm://[server-ip]:9080/[user-id-string]/banjo`
- iOS app syncs/downloads the contents of the `banjo` realm

## Expected Results

What did you expected to happen?

- iOS app/simulator is able to download contents of `banjo` realm so that I can display view controller that allows me to search over it
- when looking at the simulator's documents directory, i cannot find the `banjo` realm; granted, I'm unsure this is what I am expected to see

![File Not Found](https://github.com/jarrodparkes/images/blob/master/banjo-file-not-found.png?raw=true)

## Actual Results

What did happened instead?  

`fatal error: 'try!' expression unexpectedly raised an error: Error Domain=io.realm Code=5 "Directory at path '/Users/jarrodparkes/Library/Developer/CoreSimulator/Devices/B4854B46-966F-4EF1-8D36-32390701C371/data/Containers/Data/Application/6C798B08-FB44-45BD-8DE2-E98CFB0E9CDA/Documents/realm-object-server/6e8ba4335322f219d8cac5682457e206/realm%3A%2F%2F45.55.65.37%3A9080%2F19ebf526e022f8aec47c6c30eb8149c6%2Fbanjo' does not exist." UserInfo={Error Code=5, NSFilePath=/Users/jarrodparkes/Library/Developer/CoreSimulator/Devices/B4854B46-966F-4EF1-8D36-32390701C371/data/Containers/Data/Application/6C798B08-FB44-45BD-8DE2-E98CFB0E9CDA/Documents/realm-object-server/6e8ba4335322f219d8cac5682457e206/realm%3A%2F%2F45.55.65.37%3A9080%2F19ebf526e022f8aec47c6c30eb8149c6%2Fbanjo, Underlying=open("/Users/jarrodparkes/Library/Developer/CoreSimulator/Devices/B4854B46-966F-4EF1-8D36-32390701C371/data/Containers/Data/Application/6C798B08-FB44-45BD-8DE2-E98CFB0E9CDA/Documents/realm-object-server/6e8ba4335322f219d8cac5682457e206/realm%3A%2F%2F45.55.65.37%3A9080%2F19ebf526e022f8aec47c6c30eb8149c6%2Fbanjo") failed: No such file or directory, NSLocalizedDescription=Directory at path '/Users/jarrodparkes/Library/Developer/CoreSimulator/Devices/B4854B46-966F-4EF1-8D36-32390701C371/data/Containers/Data/Application/6C798B08-FB44-45BD-8DE2-E98CFB0E9CDA/Documents/realm-object-server/6e8ba4335322f219d8cac5682457e206/realm%3A%2F%2F45.55.65.37%3A9080%2F19ebf526e022f8aec47c6c30eb8149c6%2Fbanjo' does not exist.}: file /Library/Caches/com.apple.xbs/Sources/swiftlang/swiftlang-800.0.58.6/src/swift/stdlib/public/core/ErrorType.swift, line 178`

![Crash Image](https://github.com/jarrodparkes/images/blob/master/banjo-realm-crash.png?raw=true)

## Steps to Reproduce

What are steps we can follow to reproduce this issue?

- Try connecting to my ROS using the following code (Swift)

```swift
let username = "readonly@parkestwins.com"
let password = "readonly"

SyncUser.logIn(with: .usernamePassword(username: username, password: password, register: false), server: URL(string: "http://45.55.65.37:9080")!) { user, error in
    guard let user = user else {
        fatalError(String(describing: error))
    }
    
    DispatchQueue.main.async {
        Realm.Configuration.defaultConfiguration = Realm.Configuration(syncConfiguration: SyncConfiguration(user: user, realmURL: URL(string: "realm://45.55.65.37:9080/19ebf526e022f8aec47c6c30eb8149c6/banjo")!), readOnly: true)
        self.performSegue(withIdentifier: "login", sender: self)
    }
}
```

- Once connected, try creating Realm using the sync configuration

```swift
let configuration = toRLMConfiguration(configuration: Realm.Configuration.defaultConfiguration)
return try! RLMRealm(configuration: configuration)
```

## Code Sample

Will send email to `help@realm.io` to provide access to Xcode project.

## Version of Realm and Tooling

Realm version: realm-cocoa_2.1.0, using embedded iOS Realm frameworks (swift-3.0.1)

Xcode version: Xcode 8.1, Build version 8B62

iOS/OSX version: Mac OS X, ProductVersion: 10.12.2, BuildVersion: 16C67

Dependency manager + version: Cocoapods 1.1.1

**Comment by austinzheng** [comment]
Thanks for getting in touch with us. Can you try removing the `readOnly: true` configuration option and trying again?

Note that a local Realm being read-only is different from a user having read-only permissions for a synced Realm. We don't support local Realms being marked as read-only when used with sync; a newer release should check for this condition.

**Comment by jarrodparkes** [comment]
@austinzheng that looks like it fixed the problem of not writing the Realm file locally, but it does appear the connection gets closed quickly (preventing future syncs if the data changes)

Here's the error in the console:

```
2016-12-27 17:46:42.991 Banjo[861:11053] Sync: Connection[1]: Connection closed due to error reported by server: Permission denied (BIND, REFRESH) (206)
```

**Comment by austinzheng** [comment]
Sorry to hear that. I've asked our engineers who are more familiar with the sync server product to comment. If I hear anything from them I'll pass it along immediately.

**Comment by jarrodparkes** [comment]
@austinzheng i took a closer look at the logs and found this (notice the ** marked line **):

```
2016-12-27T23:43:14.575Z - info: sync-server: Connection[4]: Connection from 127.0.0.1:38290
2016-12-27T23:43:14.613Z - info: sync-server: Connection[4]: Received: CLIENT(protocol_version=15, client_info_size=2, client_info='{}')
2016-12-27T23:43:14.689Z - info: sync-server: Connection[4]: Session[1]: Session initiated (session_ident=1).
2016-12-27T23:43:14.689Z - info: sync-server: Connection[4]: Session[1]: Received: BIND(server_path=/19ebf526e022f8aec47c6c30eb8149c6/banjo, signed_user_token='...kmPr0TXMjZ4HAu0OhpzLUd3QQpdg==', need_file_ident_pair=1)
2016-12-27T23:43:14.690Z - info: sync-server: Connection[4]: Session[1]: Sending: ALLOC(server_file_ident=6699694228862964712, client_file_ident=4, client_file_ident_secret=3634658392737338373)
2016-12-27T23:43:14.727Z - info: sync-server: Connection[4]: Session[1]: Received: IDENT(server_file_ident=6699694228862964712, client_file_ident=4, client_file_ident_secret=3634658392737338373, scan_server_version=0, scan_client_version=0, latest_server_version=0, latest_server_session_ident=0)
** 2016-12-27T23:43:14.763Z - error: sync-server: Connection[4]: Session[1]: Permission denied (message_type='upload', path='/19ebf526e022f8aec47c6c30eb8149c6/banjo'). **
2016-12-27T23:43:14.763Z - info: sync-server: Connection[4]: Session[1]: Session terminated (session_ident=1).
2016-12-27T23:43:14.763Z - info: sync-server: Connection[4]: Sending: ERROR(error_code=206, message_size=33, try_again=0, session_ident=1)
2016-12-27T23:43:14.765Z - info: sync-server: Connection[4]: Connection closed by client: End of input
```

The `upload` being denied made me assume that, for whatever reason, my app was attempting to write something. So, I went ahead and gave the user `mayWrite` permissions for the Realm and now the connection isn't being closed — and "it works".

However, this seems to circumvent the goal (a "readonly" Realm that is synced locally to a client iOS app), and it makes me curious whether I'm trying to do something Realm wasn't designed to do?

**Comment by jarrodparkes** [comment]
@austinzheng I've looked more into the project's single view controller — the only place where Realm is mentioned, and _I'm not finding any evidence of code that results in an upload/write_. So, if it is happening, then it must be happening behind the scenes unbeknownst to me.

```swift
import UIKit
import RealmSwift

// MARK: - RealmTestVC: UITableViewController

class RealmTestVC: UITableViewController {
    
    // MARK: Property
    
    var games = List<Game>()
    var notificationToken: NotificationToken!
    var realm: Realm!
    
    // MARK: Life Cycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        setupRealm()
    }
    
    // MARK: Setup
    
    func setupUI() {
        title = "Games"
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "gameCell")
    }
    
    func setupRealm() {
        let username = "readonly@parkestwins.com"
        let password = "readonly"
        
        SyncUser.logIn(with: .usernamePassword(username: username, password: password, register: false), server: URL(string: "http://45.55.65.37:9080")!) { user, error in
            guard let user = user else {
                fatalError(String(describing: error))
            }
            
            DispatchQueue.main.async {
                // Open Realm
                var configuration = Realm.Configuration(
                    syncConfiguration: SyncConfiguration(user: user, realmURL: URL(string: "realm://45.55.65.37:9080/19ebf526e022f8aec47c6c30eb8149c6/banjo")!)                    
                )
                self.realm = try! Realm(configuration: configuration)
                
                // Show initial tasks
                func updateGames() {
                    if self.games.realm == nil, let platform = self.realm.objects(Platform.self).first {
                        self.games = platform.games
                    }
                    self.tableView.reloadData()
                }
                updateGames()
                
                // Notify us when Realm changes
                self.notificationToken = self.realm.addNotificationBlock { _ in
                    updateGames()
                }
            }
        }
    }
    
    deinit {
        notificationToken.stop()
    }
    
    // MARK: UITableViewDataSource
    
    override func tableView(_ tableView: UITableView?, numberOfRowsInSection section: Int) -> Int {
        return games.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "gameCell", for: indexPath)
        let game = games[indexPath.row]
        cell.textLabel?.text = game.title
        return cell
    }
}
```

**Comment by jarrodparkes** [comment]
UPDATE: Speaking with @TimOliver on Slack, he said part of the issue could be the sync URL. He recommended using something like `realm://localhost:9080/MyPublicRealm`, so I did.

Currently, my sync URL looks like this `realm://127.0.0.1:9080/banjo27`, but im still getting the following output in the console:
```2017-01-04 06:27:20.718 Banjo[3293:30756] Sync: Connection[1]: Resolving '127.0.0.1:9080'
2017-01-04 06:27:20.719 Banjo[3293:30756] Sync: Connection[1]: Connecting to endpoint '127.0.0.1:9080' (1/1)
2017-01-04 06:27:20.719 Banjo[3293:30756] Sync: Connection[1]: Connected to endpoint '127.0.0.1:9080' (from '127.0.0.1:51987')
2017-01-04 06:27:20.721 Banjo[3293:30756] Sync: Connection[1]: Sending: CLIENT(protocol_version=15, client_info_size=2)
2017-01-04 06:27:20.722 Banjo[3293:30756] Sync: Connection[1]: Received: ERROR(error_code=206, message_size=33, try_again=0, session_ident=1)
2017-01-04 06:27:20.722 Banjo[3293:30756] Sync: Connection[1]: Connection closed due to error reported by server: Permission denied (BIND, REFRESH) (206)
```
^ it just keeps repeating over and over as the app continues to execute. it also appears that the initial data sync happens, b/c i can access the Realm objects, but I'm unsure if any subsequent notifications are coming through without error b/c of the output I'm seeing.

**Comment by jarrodparkes** [comment]
**UPDATE**: Trying another workaround... creating a `SyncPermissionChange` object. My code isn't much different from the example on the docs:

```swift
let managementRealm = try! user.managementRealm()
try! managementRealm.write {
    let permissionChange = SyncPermissionChange(
        realmURL: realmURL.absoluteString,    // The remote Realm URL on which to apply the changes
        userID: "*", // The user ID for which these permission changes should be applied
        mayRead: true,         // Grant read access
        mayWrite: false,        // Grant write access
        mayManage: false)      // Grant management access
    managementRealm.add(permissionChange)
}
```

But this still isn't changing the output 😢 

```
2017-01-07 11:09:24.132 Banjo[6360:89938] Sync: Connection[1]: Resolving '127.0.0.1:9080'
2017-01-07 11:09:24.134 Banjo[6360:89938] Sync: Connection[1]: Connecting to endpoint '127.0.0.1:9080' (1/1)
2017-01-07 11:09:24.135 Banjo[6360:89938] Sync: Connection[1]: Connected to endpoint '127.0.0.1:9080' (from '127.0.0.1:50782')
2017-01-07 11:09:24.137 Banjo[6360:89938] Sync: Connection[1]: Sending: CLIENT(protocol_version=15, client_info_size=2)
2017-01-07 11:09:24.137 Banjo[6360:89938] Sync: Connection[1]: Received: ERROR(error_code=206, message_size=33, try_again=0, session_ident=1)
2017-01-07 11:09:24.137 Banjo[6360:89938] Sync: Connection[1]: Connection closed due to error reported by server: Permission denied (BIND, REFRESH) (206)
```

@TimOliver any thoughts?

**Comment by jarrodparkes** [comment]
Looks like this issue/bug has been reported before -- #4296

**Comment by dakeshi** [comment]
I got the same issue when I run the demo app following this posting : https://realm.io/news/track-santa-with-realm-swift-database-platform-part-3/.  The demo app generates the following message continuously:

>2017-01-10 09:10:37.601744 Santa Tracker[1679:603379] Sync: Connection[1]: Resolving '162.243.150.99:9080'
2017-01-10 09:10:37.603943 Santa Tracker[1679:603379] Sync: Connection[1]: Connecting to endpoint '162.243.150.99:9080' (1/1)
2017-01-10 09:10:37.895627 Santa Tracker[1679:603379] Sync: Connection[1]: Connected to endpoint '162.243.150.99:9080' (from '192.168.0.3:55752')
2017-01-10 09:10:38.101107 Santa Tracker[1679:603379] Sync: Connection[1]: Sending: CLIENT(protocol_version=15, client_info_size=2)
2017-01-10 09:10:38.101582 Santa Tracker[1679:603379] Sync: Connection[1]: Received: ERROR(error_code=206, message_size=33, try_again=0, session_ident=1)
2017-01-10 09:10:38.102489 Santa Tracker[1679:603379] Sync: Connection[1]: Connection closed due to error reported by server: Permission denied (BIND, REFRESH) (206)

The demo app normally shows santa's route and the number of presents with initial sync but I can not certain that it will work in the future. (@jarrodparkes pointed out the same concern in the above.)
> it also appears that the initial data sync happens, b/c i can access the Realm objects, but I'm unsure if any subsequent notifications are coming through without error b/c of the output I'm seeing.

**Comment by rmd2** [comment]
Same as dakeshi here... Any news?

**Comment by rmd2** [comment]
Same as dakeshi here... Any news?

**Comment by jarrodparkes** [comment]
@austinzheng not sure if you can raise the priority of this higher. this is actually blocking me from my first app release — if i cannot have a "readonly" sync'ed realm, and must resort to giving users mayWrite privileges (my current workaround), then it would be possible for a savvy user to complete wreak the Realm 😢 

unless there is something I'm missing?

**Comment by austinzheng** [comment]
Hi @jarrodparkes, apologies for the delayed response. I'm trying to get in touch with our engineers who'd know how to answer your question, and will post something as soon as I hear from them.

**Comment by jarrodparkes** [comment]
@austinzheng sounds good — i'm doing my best to find a workaround at this time, but I've struggled to find anything that "feels right".

**Comment by ianbradbury-bizzibody** [comment]
I am now also coming up against this issue - and my app logs look nearly identical to the examples already posted here.

`2017-01-21 06:43:51.469 Realm_Owner[14612:2381284] Sync: Connection[1]: Connection closed due to error`

I only noticed this yesterday, once I had upgraded both the client and server software to the latest "v1" (Swift / client 2.3.0 and server Ubuntu 16.04 / 1.0.0-BETA-4.11).  Previously the server was running a fairly old (nearly 2 months) release.

As I am only "playing" I am tempted to burn the server and start over.  Unless anyone has found a workaround/solution.

**Comment by jarrodparkes** [comment]
@austinzheng has there been any updates on this since the new release?

**Comment by jpsim** [comment]
Read-only sync permissions have not yet been fully implemented on the client-side, and may not work at all depending on your situation. We're tracking this internally as realm/product#195, and we have other related bugs that are opened that should be reviewed and consolidated into a single bug tracking the fact that read-only or write-only sync permissions aren't well supported.

**Comment by jpsim** [comment]
This has been fixed in #4910.

Note that if a sync user ever has the possibility of only having read access to a Realm, the only safe way to access that Realm is through the `asyncOpen` API. This is because the synchronous API provides access to the Realm immediately (yay offline-first!), which eagerly writes some metadata information to the Realm (e.g. schema), which then fails if the server rejects the changes due to insufficient permissions.

**Comment by sipersso** [comment]
I just wasted days on trying to figure this out. Documentation is nice, but I wish the error message in the console would be more helpful.
